mmtk/src/scanning.rs

```diff
@@ -29,59 +29,55 @@ pub struct VMScanning {}
 
 const DUMP_REF: bool = false;
 
-// Fix the size of any `ProcessEdgesWork` to 4096. THe constant is also hard-coded in Java code.
-pub const PROCESS_EDGES_WORK_SIZE: usize = 4096;
+// The mmtk-core no longer exposes the capacity of work packets through the API.
```

Comment:
`RootsWorkFactory` could still expose a constant for the capacity. I guess a binding should match this value to the capacity of the actual work packet impl. Othrewise, we either waste space (if binding's capacity < the work packet capacity), or need to realloc space (if binding's capacity > the work packet capacity).

This issue also exists for the OpenJDK binding.

Reply:
From my understanding, work packets do not have a "capacity".  The `ProcessEdgesBase::edges` field is initialized in `ProcessEdgesWork::new(edges: Vec<Address>, roots: bool, mmtk: &'static MMTK<Self::VM>) -> Self`.  It will accept whatever `edges` argument passed to it, no matter whether it is big or small.

But when `ScanObjects` creates other `ProcessEdgesWork` work packets, it uses the `ProcessEdgesWork::CAPACITY` constant.  However, it is part of the implementation detail of mmtk-core.  I guess this constant is what you mean by "capacity".

From my understanding, a work-packet system may have an "ideal" packet size.  It should be coarse enough to minimise the scheduling cost, but fine enough to maximise parallelism.  The ideal size may not be a constant.  Different work packets are processed at different speed, and the speed may even be affected by the Plan in use.  I am not sure if 4096 is good or bad.

The problem with JikesRVM, I think, is that the when the rust part creates a Vec buffer, it needs to communicate with the Java part about the capacity of the Vec buffer.  However, Java only allows one single return value, so we cannot return both the buffer and the capacity to Java.  In the OpenJDK binding, we have the `NewBuffer` struct type which is returned to C++.  in JikesRVM, the Java part and the Rust part have to agree upon a constant buffer size.

I can add two extra methods: `edges_buffer_size_hint(&self) -> usize` and `nodes_buffer_size_hint(&self) -> usize`, to the `RootsWorkFactory` trait.  They shall return the ideal buffer sizes.  OpenJDK can use it as a hint to create the Vec given to `NewBuffer`.  But JikesRVM may have difficulty making use of this feature.

Reply:
Yeah. That is what I meant, though I suggested using a constant in the trait. But anyway, maybe we should not overdesign this: currently we don't have any `ProcessEdgesWork` that overrides the constant, and possibly in the future we won't override it either. Furthermore, we may be able to simply make the capacity a global constant (e.g. a global `pub const CAPACITY` rather than `ProcessEdgesWork::CAPACITY`). In that case, we just need to expose that global constant. Maybe we don't need to do anything for this for now.